File: src/platform/domain/timeline.py
"""Domain models for Unified Intelligence Timeline & Explainability Boundary.

Defines timeline categories, severity levels, timeline items, and explainability payloads
with immutability and strict schema validation.
"""

from dataclasses import dataclass, field
from enum import Enum
import time
from typing import Any, Dict, List, Optional


class TimelineCategory(str, Enum):
    """Categories of intelligence timeline events."""

    MARKET = "market"
    SIGNAL = "signal"
    NOTIFICATION = "notification"
    HEALTH = "health"
    WORKSPACE = "workspace"
    MARKET_INTELLIGENCE = "market_intelligence"
    SYSTEM = "system"


class TimelineSeverity(str, Enum):
    """Severity levels for timeline items."""

    INFO = "info"
    SUCCESS = "success"
    WARNING = "warning"
    ERROR = "error"


@dataclass(frozen=True)
class TimelineItem:
    """Domain model representing a single real chronological timeline event."""

    item_id: str
    timestamp: float
    category: TimelineCategory
    severity: TimelineSeverity
    title: str
    summary: str
    source: str
    route: str
    target_user_id: Optional[str] = None
    explainable: bool = False
    payload: Dict[str, Any] = field(default_factory=dict)

    def __post_init__(self) -> None:
        if not isinstance(self.item_id, str) or not self.item_id.strip():
            raise ValueError("item_id must be a non-empty string")
        object.__setattr__(self, "item_id", self.item_id.strip())

        if not isinstance(self.timestamp, (int, float)) or self.timestamp < 0:
            raise ValueError("timestamp must be a non-negative real number")
        object.__setattr__(self, "timestamp", float(self.timestamp))

        if not isinstance(self.category, TimelineCategory):
            if isinstance(self.category, str):
                try:
                    cat = TimelineCategory(self.category.strip().lower())
                    object.__setattr__(self, "category", cat)
                except ValueError:
                    raise ValueError(f"Invalid category: {self.category}")
            else:
                raise ValueError("category must be a TimelineCategory enum or valid string")

        if not isinstance(self.severity, TimelineSeverity):
            if isinstance(self.severity, str):
                try:
                    sev = TimelineSeverity(self.severity.strip().lower())
                    object.__setattr__(self, "severity", sev)
                except ValueError:
                    raise ValueError(f"Invalid severity: {self.severity}")
            else:
                raise ValueError("severity must be a TimelineSeverity enum or valid string")

        if not isinstance(self.title, str) or not self.title.strip():
            raise ValueError("title must be a non-empty string")
        object.__setattr__(self, "title", self.title.strip())

        if not isinstance(self.summary, str) or not self.summary.strip():
            raise ValueError("summary must be a non-empty string")
        object.__setattr__(self, "summary", self.summary.strip())

        if not isinstance(self.source, str) or not self.source.strip():
            raise ValueError("source must be a non-empty string")
        object.__setattr__(self, "source", self.source.strip())

        if not isinstance(self.route, str) or not self.route.strip():
            raise ValueError("route must be a non-empty string")
        object.__setattr__(self, "route", self.route.strip())

        if self.target_user_id is not None:
            if not isinstance(self.target_user_id, str) or not self.target_user_id.strip():
                raise ValueError("target_user_id must be a non-empty string if provided")
            object.__setattr__(self, "target_user_id", self.target_user_id.strip())

        if not isinstance(self.explainable, bool):
            raise ValueError("explainable must be a boolean")

        if not isinstance(self.payload, dict):
            raise ValueError("payload must be a dict")

    def to_dict(self) -> Dict[str, Any]:
        """Convert timeline item to dictionary representation."""
        return {
            "item_id": self.item_id,
            "timestamp": self.timestamp,
            "category": self.category.value,
            "severity": self.severity.value,
            "title": self.title,
            "summary": self.summary,
            "source": self.source,
            "route": self.route,
            "target_user_id": self.target_user_id,
            "explainable": self.explainable,
            "payload": dict(self.payload),
        }


@dataclass(frozen=True)
class ExplainabilityPayload:
    """Domain model representing explainable context for permitted intelligence items."""

    item_id: str
    item_type: str
    received_at: float
    source: str
    freshness_status: str
    permitted_metadata: Dict[str, Any] = field(default_factory=dict)
    permitted_market_context: Dict[str, Any] = field(default_factory=dict)
    permitted_risk_context: Dict[str, Any] = field(default_factory=dict)
    explainability_notes: List[str] = field(default_factory=list)

    def __post_init__(self) -> None:
        if not isinstance(self.item_id, str) or not self.item_id.strip():
            raise ValueError("item_id must be a non-empty string")
        object.__setattr__(self, "item_id", self.item_id.strip())

        if not isinstance(self.item_type, str) or not self.item_type.strip():
            raise ValueError("item_type must be a non-empty string")
        object.__setattr__(self, "item_type", self.item_type.strip())

        if not isinstance(self.received_at, (int, float)) or self.received_at < 0:
            raise ValueError("received_at must be a non-negative real number")
        object.__setattr__(self, "received_at", float(self.received_at))

        if not isinstance(self.source, str) or not self.source.strip():
            raise ValueError("source must be a non-empty string")
        object.__setattr__(self, "source", self.source.strip())

        if not isinstance(self.freshness_status, str) or not self.freshness_status.strip():
            raise ValueError("freshness_status must be a non-empty string")
        object.__setattr__(self, "freshness_status", self.freshness_status.strip())

        if not isinstance(self.permitted_metadata, dict):
            raise ValueError("permitted_metadata must be a dict")

        if not isinstance(self.permitted_market_context, dict):
            raise ValueError("permitted_market_context must be a dict")

        if not isinstance(self.permitted_risk_context, dict):
            raise ValueError("permitted_risk_context must be a dict")

        if not isinstance(self.explainability_notes, list):
            raise ValueError("explainability_notes must be a list")

    def to_dict(self) -> Dict[str, Any]:
        """Convert explainability payload to dictionary representation."""
        return {
            "item_id": self.item_id,
            "item_type": self.item_type,
            "received_at": self.received_at,
            "source": self.source,
            "freshness_status": self.freshness_status,
            "permitted_metadata": dict(self.permitted_metadata),
            "permitted_market_context": dict(self.permitted_market_context),
            "permitted_risk_context": dict(self.permitted_risk_context),
            "explainability_notes": list(self.explainability_notes),
        }

File: web/src/architecture/timeline.ts
import type { HostSnapshot } from "./hostView";
import type { NotificationItem } from "./notification";

export type TimelineCategory =
  | "market"
  | "signal"
  | "notification"
  | "health"
  | "workspace"
  | "market_intelligence"
  | "system";

export type TimelineSeverity = "info" | "success" | "warning" | "error";

export interface TimelineItem {
  itemId: string;
  timestamp: number;
  formattedTime: string;
  category: TimelineCategory;
  severity: TimelineSeverity;
  title: string;
  summary: string;
  source: string;
  route: string;
  explainable: boolean;
  payload: Record<string, unknown>;
}

export interface ExplainabilityPayload {
  itemId: string;
  itemType: string;
  receivedAt: string;
  source: string;
  freshnessStatus: string;
  permittedMetadata: Record<string, unknown>;
  permittedMarketContext: Record<string, unknown>;
  permittedRiskContext: Record<string, unknown>;
  explainabilityNotes: string[];
}

export function extractTimelineFromHostSnapshot(
  snapshot: HostSnapshot,
  notifications: readonly NotificationItem[] = []
): TimelineItem[] {
  const items: TimelineItem[] = [];

  // 1. Signal Timeline Events (only active live signals, never NO SIGNAL or stale/historical)
  if (
    snapshot.signal &&
    snapshot.signal.status === "active" &&
    snapshot.signal.signalId &&
    snapshot.signal.action &&
    snapshot.signal.action !== "NO SIGNAL" &&
    snapshot.signal.action !== "STALE SIGNAL"
  ) {
    const sig = snapshot.signal;
    const action = sig.action;
    const tsStr = sig.timestamp || snapshot.generatedAt || new Date().toISOString();
    const tsVal = Date.parse(tsStr) || Date.now();

    items.push({
      itemId: `sig-${sig.signalId}`,
      timestamp: tsVal,
      formattedTime: new Date(tsVal).toUTCString(),
      category: "signal",
      severity: action === "BUY" || action === "SELL" ? "success" : "info",
      title: `${action} Signal Emitted (${snapshot.market?.symbol || "XAUUSD"})`,
      summary: sig.message || `Signal action ${action} emitted by ${sig.strategyName || "Project 1"}.`,
      source: snapshot.project1?.adapterName || "Project1GatewayAdapter",
      route: "/signals",
      explainable: true,
      payload: {
        signalId: sig.signalId,
        action,
        symbol: snapshot.market?.symbol,
        timeframe: sig.timeframe,
        confidence: sig.confidence,
        strategyName: sig.strategyName,
        status: sig.status,
        metadata: sig.metadata || {},
      },
    });
  }

  // 2. Market Events & Quote Context
  if (snapshot.market) {
    const mkt = snapshot.market;
    const tsStr = mkt.lastFetchedAt || snapshot.generatedAt || new Date().toISOString();
    const tsVal = Date.parse(tsStr) || Date.now();

    items.push({
      itemId: `mkt-${mkt.symbol.toLowerCase()}`,
      timestamp: tsVal,
      formattedTime: new Date(tsVal).toUTCString(),
      category: "market",
      severity: mkt.status === "connected" ? "info" : "warning",
      title: `Market Feed: ${mkt.symbol}`,
      summary: mkt.message || `Market data status: ${mkt.status}`,
      source: mkt.provider?.name || "MarketDataService",
      route: "/markets",
      explainable: true,
      payload: {
        symbol: mkt.symbol,
        timeframe: mkt.timeframe,
        status: mkt.status,
        quote: mkt.quote,
        provider: mkt.provider,
      },
    });
  }

  // 3. User Notifications Feed
  for (const notif of notifications) {
    let cat: TimelineCategory = "notification";
    if (notif.category === "signal") cat = "signal";
    else if (notif.category === "market_health") cat = "health";
    else if (notif.category === "workspace") cat = "workspace";

    let sev: TimelineSeverity = "info";
    if (notif.severity === "success") sev = "success";
    else if (notif.severity === "warning") sev = "warning";
    else if (notif.severity === "error") sev = "error";

    items.push({
      itemId: `notif-${notif.id}`,
      timestamp: notif.rawTimestamp,
      formattedTime: notif.timestamp,
      category: cat,
      severity: sev,
      title: notif.title,
      summary: notif.message,
      source: "NotificationCenter",
      route: "/notifications",
      explainable: false,
      payload: notif.metadata || {},
    });
  }

  // 4. Health & System Monitoring Events
  if (snapshot.monitoring) {
    const mon = snapshot.monitoring;
    const tsVal = Date.now();

    items.push({
      itemId: "health-system-mon",
      timestamp: tsVal,
      formattedTime: new Date(tsVal).toUTCString(),
      category: "health",
      severity: mon.health === "healthy" ? "success" : "warning",
      title: "Data & Health Status Check",
      summary: mon.message || "Freshness and provider readiness evaluated.",
      source: "ProviderMonitoringService",
      route: "/health",
      explainable: true,
      payload: {
        freshness: mon.freshness,
        health: mon.health,
        status: mon.status,
      },
    });
  }

  // Sort chronologically descending
  items.sort((a, b) => b.timestamp - a.timestamp);

  return items;
}

export function generateExplainabilityPayload(
  item: TimelineItem,
  snapshot: HostSnapshot
): ExplainabilityPayload {
  const permittedMetadata: Record<string, unknown> = {
    ...(item.payload || {}),
  };
  // Redact protected parameters if present in frontend payload
  delete permittedMetadata.sensitive_parameters;
  delete permittedMetadata.indicator_logic;
  delete permittedMetadata.lab_research;

  const permittedMarketContext = {
    symbol: snapshot.market?.symbol || "XAUUSD",
    timeframe: snapshot.market?.timeframe || "1h",
    status: snapshot.market?.status || "disconnected",
    provider: snapshot.market?.provider?.name || "None",
    lastPrice: snapshot.market?.quote?.last ?? snapshot.market?.quote?.mid ?? null,
  };

  const permittedRiskContext = {
    entryPrice: snapshot.risk?.entry || null,
    stopLoss: snapshot.risk?.stopLoss || null,
    takeProfits: snapshot.risk?.takeProfits || [],
    status: snapshot.risk?.status || "unavailable",
    riskRule: "1% Equity Risk Standard",
  };

  const explainabilityNotes = [
    `Context compiled from verified platform boundary '${item.source}'.`,
    "Articulates observed market quotes, provider states, and emitted signal metadata.",
    "Project 1 proprietary strategy math, indicators, and lab parameters remain protected.",
  ];

  if (!snapshot.security?.isAdmin) {
    explainabilityNotes.push(
      "Standard User Profile Active: Internal parameter weights and lab research are filtered."
    );
  }

  return {
    itemId: item.itemId,
    itemType: item.category,
    receivedAt: item.formattedTime,
    source: item.source,
    freshnessStatus: snapshot.monitoring?.freshness || "fresh",
    permittedMetadata,
    permittedMarketContext,
    permittedRiskContext,
    explainabilityNotes,
  };
}
